Buffer abstractions

pearpanda opened the pull request:
The following (light) abstractions have been implemented:
- Vertex Buffer
- Index Buffer
- Vertex Array

Commit:
Buffer abstractions

The following (light) abstractions have been implemented:
- Vertex Buffer
- Index Buffer
- Vertex Array

luka-j commented on src/renderer/buffer/VertexArray.cpp:
```diff
@@ -0,0 +1,56 @@
+    glBindVertexArray(0);
+}
+
+static unsigned int id(const ElementType& type);
```
malo mi bode oči ova deklaracija u sred fajla :D

luka-j commented on src/renderer/buffer/VertexArray.cpp:
```diff
@@ -0,0 +1,56 @@
+    }
+}
+
+static unsigned int id(const ElementType& type) {
```
možda grešim ali ovo deluje kao da može da pripada onom util namespace-u gde su i druge ElementType fje (sa malo deskriptivnijim imenom), i onda ovde samo da se uključi i taj header ovde

pearpanda replied:
Mislim da bi možda trebalo ubaciti kao metodu `ElementType` strukture. Sigurno ima više smisla.

Moje razmišljanje je bilo da ovim samo VertexArray zavisi od konkretne grafičke biblioteke (OpenGL, DirectX), ali mislim da za tim nema potrebe, fokusiramo se na OpenGL.

pearpanda replied:
`ElementType` je `enum class`... Da, treba ubaciti u namespace `util`.

luka-j commented on include/rg/renderer/buffer/IndexBuffer.hpp:
```diff
@@ -0,0 +1,24 @@
+    [[nodiscard]] unsigned int count() const;
+
+private:
+    unsigned int bufferID;
```
trebalo bi da se dogovorimo oko neke konvencije pre nego što ovo poraste—ja sam pratio [Google Style Guide](https://google.github.io/styleguide/cppguide.html#Variable_Names) (uglavnom) pa sam imenovao promenljive i gettere/settere u snake case-u; vidim da i ti imaš npr VertexArray::get_elements. Ako ti je lepše i konzistentnije, možemo da sve stavimo u camel case pa da bude Java-style (ja svakako najviše s Javom radim, nego znam da ima ljudi koji ne vole :D), samo onda ispravi onaj getter. Either way, ne bih ostavljao ovo `ID`, nema nikakvo specijalno značenje niti je akronim: može i samo bufferId.

pearpanda replied:
Slažem se.
Pošto smo krenuli sa nekim Google C++ stilom, predlažem da nastavimo sa tim.
Doduše, imena svih članova klasa onda treba promeniti tako da se završavaju donjom crtom (_).

luka-j commented on include/rg/renderer/buffer/VertexArray.hpp:
```diff
@@ -0,0 +1,22 @@
+    void unbind() const;
+    void recordLayout(const VertexBuffer& vb, const VertexLayout& layout) const;
+private:
+    unsigned int arrayID;
```
naming

luka-j commented on include/rg/renderer/buffer/VertexBuffer.hpp:
```diff
@@ -0,0 +1,19 @@
+    void unbind() const;
+
+private:
+    unsigned int bufferID;
```
naming

luka-j commented on include/rg/renderer/buffer/VertexLayout.hpp:
```diff
@@ -0,0 +1,47 @@
+namespace util {
+
+constexpr std::size_t size(const ElementType& type);
+LayoutElement floatVector(unsigned int count);
+LayoutElement uintVector(unsigned int count);
```
Mene ove dve metode dosta podsećaju na [named constructor idiom](https://en.wikibooks.org/wiki/More_C%2B%2B_Idioms/Named_Constructor). Pored ove dve, postoje još samo dve nepokrivene varijante gde je normalized=true: mislim da nije veliko cimanje dodati još dve metode (npr. floatVectorNormalized & uintVectorNormalized) i staviti da je konstruktor private. Ako misliš da ćemo dodavati još neka polja u LayoutElement onda ovo nije baš održivo, pa bi trebalo koristiti builder pattern ili prosto ostaviti public konstruktor; svakako, ovo mi više liči kao da "pripada" LayoutElement-u kao static metoda nego negde spolja (koji god da je visibility konstruktora).

pearpanda replied:
Moje razmišljanje je da se od svih mogućih kombinacija najčešće pojavljuje kombinacija nenormalizovani float x n elemenata. Zbog toga sam želeo da napravim neku skraćenicu za tu najčešću varijantu, mada dok se objasni šta je i nije baš neka skraćenica.

Smatram da nema potrebe za definisanje svih kombinacija. Iz mog objašnjenja je ideja otprilike tu zašto ovo postoji, i proširenja u `ElementType` zaista ne utiču na taj najčešći slučaj.

Dodatno, trenutni tipovi koje OpenGL podržava ne deluju previše korišćeni, izuzev `float` i retko `unsigned int`. Dodavanje novih tipova u `ElementType` je veoma malo verovatno, zahtevalo bi proširenje tipova koje podržava OpenGL.

Sa tim da može da se doda kao static metoda se slažem, mada mi je nešto ovog tipa delovalo više kao neka utility fukncija kada sam pisao kod. Ako si za, možemo da prebacimo `floatVector` sa tim, ili nekim sličnim nazivom kao static metodu u `LayoutElement`, a slučaj `unsigned int` bismo mogli čak i da obrišemo.

luka-j commented on include/rg/renderer/buffer/VertexLayout.hpp:
```diff
@@ -0,0 +1,47 @@
+
+} // namespace util
+
+struct LayoutElement {
```
neki poseban razlog što je ovo struct a ne class?

pearpanda replied:
Sva polja se koriste, i ne postoje restrikcije njihovog odnosa.

Jednostavnost koda.

Takođe, ova "struktura" će se koristiti samo jednom za svaki atribut vertex buffer-a, i verovatno će se dodavati usput u VertexLayout ili će biti strpano u jedan vektor. Nakon toga, vezivanjem preko vertex array-a, završava se uloga tih konkretnih instanci. Nije mi delovalo vredno klase.

luka-j commented on include/rg/renderer/buffer/VertexLayout.hpp:
```diff
@@ -0,0 +1,47 @@
+
+    void push(LayoutElement element);
+    [[nodiscard]] std::vector<unsigned int> offsets() const;
+    [[nodiscard]] unsigned int stride() const;
```
ne bi bilo loše naznačiti da je ovo getter u imenu, posebno zato što je `stride` i glagol (ovo const jeste hint, mada ne _toliko_ očigledan)

pearpanda replied:
OK, može.

luka-j commented on include/rg/renderer/buffer/VertexLayout.hpp:
```diff
@@ -0,0 +1,47 @@
+} // namespace util
+
+struct LayoutElement {
+    ElementType type;
+    unsigned int count;
+    bool normalized;
```
ne sviđa mi se baš što su ovo public mutable polja (mogu da se menjaju od spolja — ima li potrebe za tim?)

pearpanda replied:
Moje razmišljanje je kontra, nema potrebe za zaštitom ovih polja od menjanja. Štaviše, očekujem da instance ovog tipa odmah budu prosleđene nekoj metodi ili vektoru ili nešto slično, pa ne bi se ni menjali. Možda nisam u pravu, doduše.

Commit:
Move and rename `id` to `intValue`

The function is now defined with other VertexLayout utility functions,
in the `rg::util` namespace.

In the process, discovered and fixed error pertaining to `constexpr`
funcitons. Namely, they do not link, since they are implicitly `inline`.
As a result, the `constexpr` specifier was removed from the
`rg::util::size` function.

Commit:
Fix naming convention: Google C++ style

Commit:
Replace `floatVector` with new implementation

Functions `floatVector` and `uintVector` have been removed.

A new header and source file have been added to the util group,
containing functions which generate commonly used layout elements
(float, 2D vector, 3D vector and 4D vector).
